usr/src/man/man1/ckuid.1
'\" te
.\"  Copyright 1989 AT&T  Copyright (c) 1992, Sun Microsystems, Inc.  All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH CKUID 1 "Sep 14, 1992"
.SH NAME
ckuid, erruid, helpuid, valuid \- prompts for and validates a user ID
.SH SYNOPSIS
.LP
.nf
\fBckuid\fR [\fB-Q\fR] [\fB-W\fR \fIwidth\fR] [\fB-m\fR] [\fB-d\fR \fIdefault\fR] [\fB-h\fR \fIhelp\fR]
     [\fB-e\fR \fIerror\fR] [\fB-p\fR \fIprompt\fR] [\fB-k\fR \fIpid\fR [\fB-s\fR \fIsignal\fR]]
.fi

.LP
.nf
\fB/usr/sadm/bin/erruid\fR [\fB-W\fR \fIwidth\fR] [\fB-e\fR \fIerror\fR]
.fi

.LP
.nf
\fB/usr/sadm/bin/helpuid\fR [\fB-W\fR \fIwidth\fR] [\fB-m\fR] [\fB-h\fR \fIhelp\fR]
.fi

.LP
.nf
\fB/usr/sadm/bin/valuid\fR \fIinput\fR
.fi

.SH DESCRIPTION
.sp
.LP
The \fBckuid\fR utility prompts a user and validates the response. It defines,
among other things, a prompt message whose response should be an existing user
\fBID,\fR text for help and error messages, and a default value (which are
returned if the user responds with a RETURN).
.sp
.LP
All messages are limited in length to 70 characters and are formatted
automatically. Any white space used in the definition (including NEWLINE) is
stripped. The \fB-W\fR option cancels the automatic formatting. When a tilde is
placed at the beginning or end of a message definition, the default text is
inserted at that point, allowing both custom text and the default text to be
displayed.
.sp
.LP
If the prompt, help or error message is not defined, the default message (as
defined under \fBNOTES\fR) is displayed.
.sp
.LP
Three visual tool modules are linked to the \fBckuid\fR command. They are
\fBerruid\fR (which formats and displays an error message), \fBhelpuid\fR
(which formats and displays a help message), and \fBvaluid\fR (which validates
a response). These modules should be used in conjunction with  \fBFML\fR
objects. In this instance, the  \fBFML\fR object defines the prompt.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-d\fR\fI default\fR\fR
.ad
.RS 14n
Defines the default value as \fIdefault\fR. The default is not validated and so
does not have to meet any criteria.
.RE

.sp
.ne 2
.na
\fB\fB-e\fR \fIerror\fR\fR
.ad
.RS 14n
Defines the error message as \fIerror\fR.
.RE

.sp
.ne 2
.na
\fB\fB-h\fR \fIhelp\fR\fR
.ad
.RS 14n
Defines the help messages as \fIhelp\fR.
.RE

.sp
.ne 2
.na
\fB\fB-k\fR\fI pid\fR\fR
.ad
.RS 14n
Specifies that process  \fBID\fR \fIpid\fR is to be sent a signal if the user
chooses to abort.
.RE

.sp
.ne 2
.na
\fB\fB-m\fR\fR
.ad
.RS 14n
Displays a list of all logins when help is requested or when the user makes an
error.
.RE

.sp
.ne 2
.na
\fB\fB-p\fR\fI prompt\fR\fR
.ad
.RS 14n
Defines the prompt message as \fIprompt\fR.
.RE

.sp
.ne 2
.na
\fB\fB-Q\fR\fR
.ad
.RS 14n
Specifies that quit will not be allowed as a valid response.
.RE

.sp
.ne 2
.na
\fB\fB-s\fR \fIsignal\fR\fR
.ad
.RS 14n
Specifies that the process  \fBID\fR \fIpid\fR defined with the \fB-k\fR option
is to be sent signal \fIsignal\fR when quit is chosen. If no signal is
specified, \fBSIGTERM\fR is used.
.RE

.sp
.ne 2
.na
\fB\fB-W\fR\fI width\fR\fR
.ad
.RS 14n
Specifies that prompt, help and error messages will be formatted to a line
length of \fIwidth\fR.
.RE

.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.na
\fB\fIinput\fR\fR
.ad
.RS 9n
Input to be verified against \fB/etc/passwd\fR.
.RE

.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 5n
Successful execution.
.RE

.sp
.ne 2
.na
\fB\fB1\fR\fR
.ad
.RS 5n
\fBEOF\fR on input, or negative width on  \fB-W\fR option,  or usage error.
.RE

.sp
.ne 2
.na
\fB\fB2\fR\fR
.ad
.RS 5n
Usage error.
.RE

.sp
.ne 2
.na
\fB\fB3\fR\fR
.ad
.RS 5n
User termination (quit).
.RE

.SH SEE ALSO
.sp
.LP
.BR attributes (7)
.SH NOTES
.sp
.LP
The default prompt for \fBckuid\fR is:
.sp
.in +2
.nf
Enter the login name of an existing user [?,q]:
.fi
.in -2
.sp

.sp
.LP
The default error message is:
.sp
.in +2
.nf
ERROR - Please enter the login name of an existing user.
.fi
.in -2
.sp

.sp
.LP
If the \fB-m\fR option is used, the default error message is:
.sp
.in +2
.nf
ERROR:  Please enter one of the following login names:  <List>
.fi
.in -2
.sp

.sp
.LP
The default help message is:
.sp
.in +2
.nf
Please enter the login name of an existing user.
.fi
.in -2
.sp

.sp
.LP
If the \fB-m\fR option is used, the default help message is:
.sp
.in +2
.nf
Please enter one of the following login names:  <List>
.fi
.in -2
.sp

.sp
.LP
When the quit option is chosen (and allowed), \fBq\fR is returned along with
the return code \fB3\fR. The \fBvaluid\fR module will not produce any output.
It returns  \fB0\fR for success and non-zero for failure.


usr/src/lib/libadm/common/ckuid.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*
 * Copyright (c) 1997, by Sun Microsystems, Inc.
 * All rights reserved.
 */

#pragma	ident	"%Z%%M%	%I%	%E% SMI"	/* SVr4.0 1.2 */
/*LINTLIBRARY*/

#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <pwd.h>
#include <stdlib.h>
#include <limits.h>
#include "libadm.h"

#define	PROMPT	"Enter the login name of an existing user"
#define	MESG	"Please enter the login name of an existing user."
#define	ALTMESG	"Please enter one of the following login names:\\n\\t"
#define	MALSIZ	64

#define	DELIM1 '/'
#define	BLANK ' '

static char *
setmsg(int disp)
{
	struct passwd
		*pwdptr;
	int	count;
	size_t	n, m;
	char	*msg;

	if (disp == 0)
		return (MESG);

	m = MALSIZ;
	n = sizeof (ALTMESG);
	msg = (char *) calloc(m, sizeof (char));
	(void) strcpy(msg, ALTMESG);

	setpwent();
	count = 0;
	while (pwdptr = getpwent()) {
		n += strlen(pwdptr->pw_name) + 2;
		while (n >= m) {
			m += MALSIZ;
			msg = (char *) realloc(msg, m*sizeof (char));
		}
		if (count++)
			(void) strcat(msg, ", ");
		(void) strcat(msg, pwdptr->pw_name);
	}
	endpwent();
	return (msg);
}

int
ckuid_dsp(void)
{
	struct passwd *pwdptr;

	/* if display flag is set, then list out passwd file */
	if (ckpwdfile() == 1)
		return (1);
	setpwent();
	while (pwdptr = getpwent())
		(void) printf("%s\n", pwdptr->pw_name);
	endpwent();
	return (0);
}

int
ckuid_val(char *usrnm)
{
	int	valid;

	setpwent();
	valid = (getpwnam(usrnm) ? 0 : 1);
	endpwent();
	return (valid);
}

int
ckpwdfile(void) /* check to see if passwd file there */
{
	struct passwd *pwdptr;

	setpwent();
	pwdptr = getpwent();
	if (!pwdptr) {
		endpwent();
		return (1);
	}
	endpwent();
	return (0);
}

void
ckuid_err(short disp, char *error)
{
	char	*msg;

	msg = setmsg(disp);
	puterror(stdout, msg, error);
	if (disp)
		free(msg);
}

void
ckuid_hlp(int disp, char *help)
{
	char	*msg;

	msg = setmsg(disp);
	puthelp(stdout, msg, help);
	if (disp)
		free(msg);
}

int
ckuid(char *uid, short disp, char *defstr, char *error, char *help,
	char *prompt)
{
	char	*defmesg,
		input[MAX_INPUT];

	defmesg = NULL;
	if (!prompt)
		prompt = PROMPT;

start:
	putprmpt(stderr, prompt, NULL, defstr);
	if (getinput(input)) {
		if (disp && defmesg)
			free(defmesg);
		return (1);
	}

	if (!strlen(input)) {
		if (defstr) {
			if (disp && defmesg)
				free(defmesg);
			(void) strcpy(uid, defstr);
			return (0);
		}
		if (!defmesg)
			defmesg = setmsg(disp);
		puterror(stderr, defmesg, error);
		goto start;
	} else if (strcmp(input, "?") == 0) {
		if (!defmesg)
			defmesg = setmsg(disp);
		puthelp(stderr, defmesg, help);
		goto start;
	} else if (ckquit && (strcmp(input, "q") == 0)) {
		if (disp && defmesg)
			free(defmesg);
		return (3);
	} else if (ckuid_val(input)) {
		if (!defmesg)
			defmesg = setmsg(disp);
		puterror(stderr, defmesg, error);
		goto start;
	}
	(void) strcpy(uid, input);
	if (disp && defmesg)
		free(defmesg);
	return (0);
}
